SI-6493 more robust symbol handling in existentialTransform

adriaanm opened the pull request:
To illustrate the problem, let's compute the result type for

```
def foo = { class Foo { class Bar { val b = 2 }}; val f = new Foo; new f.Bar }
```

Note the inconsistent symbol ids in:

rawSyms     = List(value f#49740, class Foo#49739, class Bar#49742)
quantifiers + their infos =
  type f.type#49774 <: AnyRef{type Bar#49762 <: AnyRef{val b#49764: Int}} with Singleton,
  type Foo#49775    <: AnyRef{type Bar#49770 <: AnyRef{val b#49772: Int}},
                              type Bar#49773 <: AnyRef{val b#49758: Int}

It makes sense to subst the symbols in tp, but the modifyInfo will have
little effect on quantifiers. Luckily, coevolveSym will fix up some of
the inconsistencies, so that we get `f.type#49774.Bar#49762`, which is
extrapolated to `AnyRef{type Bar#49868 <: AnyRef{val b: Int}}#Bar#49868`.

The intermediate type `f.type#49774.Bar#49762` motivates the need for
the `.substSym` above, as `.subst` does not touch the `Bar#49742` in
the `f#49740.Bar#49742` that we start out with.

The remaining problem is that ExistentialExtrapolation will not consider
type Bar#49868 as existentially bound. Instead, it's looking for the
stale version in quantifiers, Bar#49773. If it was looking for Bar#49868,
the result type would simply be `AnyRef{val b: Int}`

TODO: figure out how to compute the full set of quantifiers for extrapolation
      the `doSubst(tp)` will clone more symbols, so we'll probably need to
      traverse its result and somehow correlate the contained symbols to
      those in `quantifiers`...

adriaanm committed:
SI-6493 more robust symbol handling in existentialTransform

The fix is only enabled under -Xexperimental, because the
full fix is blocked by SI-2071.

To illustrate the problem, let's compute the result type for

```
def foo = { class Foo { class Bar { val b = 2 }}; val f = new Foo; new f.Bar }
```

Note the inconsistent symbol ids in:

rawSyms     = List(value f#49740, class Foo#49739, class Bar#49742)
quantifiers + their infos =
  type f.type#49774 <: AnyRef{type Bar#49762 <: AnyRef{val b#49764: Int}} with Singleton,
  type Foo#49775    <: AnyRef{type Bar#49770 <: AnyRef{val b#49772: Int}},
                              type Bar#49773 <: AnyRef{val b#49758: Int}

It makes sense to subst the symbols in tp, but the modifyInfo will have
little effect on quantifiers. Luckily, coevolveSym will fix up some of
the inconsistencies, so that we get `f.type#49774.Bar#49762`, which is
extrapolated to `AnyRef{type Bar#49868 <: AnyRef{val b: Int}}#Bar#49868`.

The intermediate type `f.type#49774.Bar#49762` motivates the need for
the `.substSym` above, as `.subst` does not touch the `Bar#49742` in
the `f#49740.Bar#49742` that we start out with.

The remaining problem is that ExistentialExtrapolation will not consider
type Bar#49868 as existentially bound. Instead, it's looking for the
stale version in quantifiers, Bar#49773. If it was looking for Bar#49868,
the result type would simply be `AnyRef{val b: Int}`

TODO: figure out how to compute the full set of quantifiers for extrapolation
      the `doSubst(tp)` will clone more symbols, so we'll probably need to
      traverse its result and somehow correlate the contained symbols to
      those in `quantifiers`...

retronym commented:
```
BUILD FAILED
/localhome/jenkins/c/workspace/pr-scala-test/scala/build.xml:1434: The following error occurred while executing this line:
/localhome/jenkins/c/workspace/pr-scala-test/scala/build-ant-macros.xml:508: Test suite finished with 3 cases failing:
fail - run/existentials3-new.scala  [compilation failed]% scalac existentials3-new.scala
existentials3-new.scala:46: error: exception during macro expansion: java.lang.StackOverflowError
    m2(f1)
      ^
existentials3-new.scala:53: error: exception during macro expansion: java.lang.StackOverflowError
    m2(f8)
      ^
existentials3-new.scala:54: error: exception during macro expansion: java.lang.StackOverflowError
    m2(f9)
      ^
existentials3-new.scala:58: error: exception during macro expansion: java.lang.StackOverflowError
    m2(g1)
      ^
existentials3-new.scala:65: error: exception during macro expansion: java.lang.StackOverflowError
    m2(g8)
      ^
existentials3-new.scala:66: error: exception during macro expansion: java.lang.StackOverflowError
    m2(g9)
      ^
6 errors found

fail - run/t1195-new.scala  [compilation failed]% scalac t1195-new.scala
t1195-new.scala:17: error: exception during macro expansion: java.lang.StackOverflowError
    m(f)
     ^
t1195-new.scala:20: error: exception during macro expansion: java.lang.StackOverflowError
    m(f1)
     ^
two errors found

fail - run/constrained-types.scala  [output differs]% scalac constrained-types.scala
% /home/jenkins/apps/jdk1.6.0_33-x64/jre/bin/java \
  -classpath \
  /localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/run/constrained-types-run.obj:/localhome/jenkins/c/workspace/pr-scala-test/scala/build/pack/lib/scala-library.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/build/pack/lib/scala-reflect.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/build/pack/lib/scala-compiler.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/build/pack/lib/scalap.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/build/pack/lib/scala-actors.jar:/localhome/jenkins/c/workspace/pr-scala-test/m2repo/org/scala-lang/modules/scala-partest_2.11.0-M8/1.0.0-RC8/scala-partest_2.11.0-M8-1.0.0-RC8.jar:/localhome/jenkins/c/workspace/pr-scala-test/m2repo/org/apache/ant/ant/1.8.4/ant-1.8.4.jar:/localhome/jenkins/c/workspace/pr-scala-test/m2repo/org/apache/ant/ant-launcher/1.8.4/ant-launcher-1.8.4.jar:/localhome/jenkins/c/workspace/pr-scala-test/m2repo/com/googlecode/java-diff-utils/diffutils/1.3.0/diffutils-1.3.0.jar:/localhome/jenkins/c/workspace/pr-scala-test/m2repo/org/scala-sbt/test-interface/1.0/test-interface-1.0.jar:/localhome/jenkins/c/workspace/pr-scala-test/m2repo/org/scala-lang/modules/scala-xml_2.11.0-M8/1.0.0-RC7/scala-xml_2.11.0-M8-1.0.0-RC7.jar:/localhome/jenkins/c/workspace/pr-scala-test/m2repo/org/scala-lang/modules/scala-parser-combinators_2.11.0-M8/1.0.0-RC5/scala-parser-combinators_2.11.0-M8-1.0.0-RC5.jar:/localhome/jenkins/c/workspace/pr-scala-test/m2repo/org/scalacheck/scalacheck_2.11.0-M8/1.11.3/scalacheck_2.11.0-M8-1.11.3.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/build/pack/lib/scala-partest-extras.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/build/pack/lib/scala-partest-javaagent.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/lib/annotations.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/lib/enums.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/lib/genericNest.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/lib/jsoup-1.3.1.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/lib/methvsfield.jar:/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/lib/nest.jar \
  -Dfile.encoding=UTF-8 \
  -Djava.library.path=/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/run \
  -Dpartest.output=/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/run/constrained-types-run.obj \
  -Dpartest.lib=/localhome/jenkins/c/workspace/pr-scala-test/scala/build/pack/lib/scala-library.jar \
  -Dpartest.reflect=/localhome/jenkins/c/workspace/pr-scala-test/scala/build/pack/lib/scala-reflect.jar \
  -Dpartest.cwd=/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/run \
  -Dpartest.test-path=/localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/run/constrained-types.scala \
  -Dpartest.testname=constrained-types \
  -Djavacmd=/home/jenkins/apps/jdk1.6.0_33-x64/jre/bin/java \
  -Djavaccmd=javac \
  -Duser.language=en \
  -Duser.country=US \
  scala.tools.nsc.MainGenericRunner \
  -usejavacp \
  Test \
  jvm > constrained-types-run.log
% diff /localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/run/constrained-types-run.log /localhome/jenkins/c/workspace/pr-scala-test/scala/test/files/run/constrained-types.check
@@ -99,7 +99,7 @@ scala> def m = {
   y
 } // x should not escape the local scope with a narrow type
 warning: there were 1 feature warning(s); re-run with -feature for details
-m: String @Annot(x.type) forSome { val x: String }
+m: String @Annot(x) forSome { val x: String }

 scala> 
``
```

xeno-by committed:
SI-8321 whitebox bundles are now recognized as such

whitebox.Context <: blackbox.Context, so in order to check for blackboxity
it's not enough to check whether the context used is <: blackbox.Context.

xeno-by committed:
minor code cleanup in the macro engine

xeno-by committed:
more tests for macro bundles

Given the recent glaring oversight in macro bundles, I have to have more
tests in order to make sure that things are going to work as they should.

xeno-by committed:
bundles now reject invalid context types

Vanilla macros only allow blackbox.Context, whitebox.Context and
PrefixType refinements thereof. Bundles should behave in the same way.

xeno-by committed:
prohibits polymorphic bundles

It's not like they were inducing bugs, but I can't see how polymorphism
can be useful for macro bundles, hence imho it's better to reduce the
number of degrees of freedom of the system.

xeno-by committed:
more helpful bundle error messages

At the moment, bundle selection mechanism is pretty picky. If a candidate
bundle's parameter isn't either blackbox.Context, whitebox.Context or
PrefixType refinement thereof, then it's not a bundle and the user
will get a generic error.

However we can be a bit more helpful and admit classes that are almost
like bundles (looksLikeMacroBundleType), have them fail isMacroBundleType,
and then emit a much prettier error message to the user that would tell them
that bundles must be monomorphic and their sole parameter should not just
be any subtype of blackbox.Context or whitebox.Context.

retronym committed:
SI-8197 Only consider immediate params for defaults, overloading

A recent commit fixed the behaviour of overloading with regards
to discarding candiates that include default arguments. The old
check was checking the wrong flag.

But, the new code is actually checking all parameter lists for
defaults, which led to a regression in scala-io, which is distilled
in the enclosed test case. Applications are typechecked one
parameter list at a time, so a holistic check for defaults doesn't
seem to make sense; only defaults in the first parameter list
ought to count.

xeno-by committed:
more clean up in the macro engine

Now when SI-7507 is fixed in starr, we can actually remove a workaround
and make a 10 line reduction in the size of Resolvers.scala.

retronym committed:
SI-8324 Fix regression in override checks for sealed classes

adeffda25 changed `Symbol#isEffectivelyFinal` to help the optimizer
by inferring finality within sealed class hierarchies.

However, this change wasn't neccesarily welcome for other clients of
that method. In the enclosed test case, we see that overriding checks
in `RefChecks` regressed.

This commit moves the enhanced version into a new method and
selectively uses it in the optimizer (and the tail call optimizer).

greenrd committed:
SI-7962 Scalac runner does not work within Emacs's terminal

  - Always set the env.emacs system property - scalac only cares
    about whether the system property has a non-empty value, not
    whether it is set or not. Fixes 7962

adriaanm committed:
Merge pull request #3572 from retronym/ticket/8197

SI-8197 Only consider immediate params for defaults, overloading

gkossakowski committed:
Merge pull request #3573 from retronym/ticket/8324

SI-8324 Fix regression in override checks for sealed classes

gkossakowski committed:
Merge pull request #3575 from retronym/ticket/scaladoc-qbin

Fix ./build/<stage>/bin/scaladoc

adriaanm committed:
Merge pull request #3574 from greenrd/SI-7962

SI-7962 Scalac runner does not work within Emacs's terminal

adriaanm committed:
Merge pull request #3571 from xeno-by/ticket/8321

SI-8321 bundles can't be whitebox
